### PR TITLE
[Clang] Implement CWG2598: Union of non-literal types

### LINKED ISSUE #77924: [clang] Union of a non-literal type and a literal type should be literal
Clang fails to compile this code : 

https://godbolt.org/z/qvxGcPx6x

```cpp
struct A {
    A();
};

template <class T>
struct opt
{
    union Data {
        
        constexpr Data() : x{} {}
        constexpr ~Data() {}

        char x;
        T data;
    };

    constexpr opt() : data{} {}
    constexpr ~opt() { if (engaged) data.data.~T(); }

    Data data;
    bool engaged = false;
};

consteval void foo()
{
    opt<A> a;
}

int main()
{
    foo();
}
```

With the error `non-literal type 'opt<A>' cannot be used in a constant expression` (in C++23, in C++20 the message is slightly different). `opt<A>` is not considered a literal type because of the implementation of CXXRecordDecl::isLiteral : 
https://github.com/llvm/llvm-project/blob/ae5d63924a6214154194c286a13c6ae74d31c086/clang/include/clang/AST/DeclCXX.h#L1457

which does not take into account that the CXXRecordDecl in question might be an union, in which case the type is literal if at least one of its member is literal. Perhaps the meaning of the bitfield `hasNonLiteralTypeFieldsOrBases` should change when the Decl is an union? since it does not seems to serve any other purposes than to detect whether the type is literal or not. 

(this causes problem with the constexpr-isation of `std::format` which i'm currently working on)

### PR DESCRIPTION
**Pull request by cor3ntin**
A union is considered a literal type unless it has no non-literal member.

This resolves CWG2096 (which makes unions with literal members literal) and CWG2598 (empty unions are literal types).

Fixes #77924

### CODE REVIEW

File: clang/lib/AST/DeclCXX.cpp [hunk 1]
```diff
@@ -1383,6 +1383,34 @@ void CXXRecordDecl::addedMember(Decl *D) {
...
+  if (hasNonLiteralTypeFieldsOrBases()) {
+    // CWG2598
+    // is an aggregate union type that has either no variant
+    // members or at least one variant member of non-volatile literal type,
+    if (!isUnion())
+      return false;
+    bool HasAtLeastOneTrivialMember =
+        fields().empty() || any_of(fields(), [this](const FieldDecl *D) {
+          return !D->getType().isVolatileQualified() &&
+                 D->getType().isTrivialType(getASTContext());
```
Comment (MitalAshok) [inline clang/lib/AST/DeclCXX.cpp:1404]: What about literal non-trivial types (e.g. `struct Literal { constexpr Literal() {} }; union union6 { NonLiteral NL; Literal L; };`?

`.isLiteralType` should work here

Comment (cor3ntin) [reply clang/lib/AST/DeclCXX.cpp:1404]: Autocompletion blunder, Nice catch!

**Review by shafik — COMMENTED** [review]
Is this a potentially breaking change since we have expanded what types are considered literals and already existing code may observe this fix?

File: clang/lib/AST/DeclCXX.cpp [hunk 2]
```diff
@@ -1383,6 +1383,34 @@ void CXXRecordDecl::addedMember(Decl *D) {
...
+  if (isLambda() && !LangOpts.CPlusPlus17)
+    return false;
+
+  if (hasNonLiteralTypeFieldsOrBases()) {
+    // CWG2598
+    // is an aggregate union type that has either no variant
+    // members or at least one variant member of non-volatile literal type,
+    if (!isUnion())
+      return false;
+    bool HasAtLeastOneTrivialMember =
```
Comment (shafik) [inline clang/lib/AST/DeclCXX.cpp:1401]: Why use trivial here? The DRs don't refer to trivial it refers to non-volatile literal type.

File: clang/test/CXX/drs/dr25xx.cpp [hunk 3]
```diff
@@ -208,3 +208,54 @@ namespace dr2565 { // dr2565: 16 open
...
+static_assert(!__is_literal(union3), "");
+
+union union4 { union4(); };
+static_assert(!__is_literal(union4), "");
+
+union union5 { static NonLiteral NL; };
+static_assert(__is_literal(union5), "");
+
+struct Literal { constexpr Literal() {} };
+union union6 { NonLiteral NL; Literal L; };
```
Comment (shafik) [inline clang/test/CXX/drs/dr25xx.cpp:258]: Can we also include the exact example from CWG2598 as well, just to be complete.

Comment (shafik) [reply clang/test/CXX/drs/dr25xx.cpp:258]: Also should we have a `static_assert` after this line?

Comment (shafik) [reply clang/test/CXX/drs/dr25xx.cpp:258]: I also noticed that gcc does not seem to support `__is_literal` only `__is_literal_type`

**Review by shafik — COMMENTED** [review]
Mostly looks good but I would like a second set of eyes.

File: clang/lib/AST/DeclCXX.cpp (hunk 2 above)
Comment (cor3ntin) [reply clang/lib/AST/DeclCXX.cpp]: Yes, that makes no sense

File: clang/test/CXX/drs/dr25xx.cpp (hunk 3 above)
Comment (cor3ntin) [reply clang/test/CXX/drs/dr25xx.cpp:258]: Yup, both are synnym in clang

**Comment by cor3ntin** [comment]
> Is this a potentially breaking change since we have expanded what types are considered literals and already existing code may observe this fix?

Some expression will become constant expressions and presumably that can be SFINAE upon, but given that this is a DR, I don't think we are affecting conformance and I don't think we make strong sfinae guarantee.

GCC is conforming with that DR in all language mode since 10.0 
https://compiler-explorer.com/z/hKnThPx1T

File: clang/lib/AST/DeclCXX.cpp [hunk 4]
```diff
@@ -1383,6 +1383,34 @@ void CXXRecordDecl::addedMember(Decl *D) {
...
+    bool HasAtLeastOneLiteralMember =
+        fields().empty() || any_of(fields(), [this](const FieldDecl *D) {
+          return !D->getType().isVolatileQualified() &&
+                 D->getType()->isLiteralType(getASTContext());
+        });
+    if (!HasAtLeastOneLiteralMember)
+      return false;
+  }
+
+  return isAggregate() || isLambda() || hasConstexprNonCopyMoveConstructor() ||
```
Comment (erichkeane) [inline clang/lib/AST/DeclCXX.cpp:1410]: Should these conditions be moved 'above' the union stuff?  I would think they are 'cheaper'.  Also, I see the 'isLambda' here, despite it only being in C++17 mode on line 1392.

File: clang/lib/AST/DeclCXX.cpp [hunk 5]
```diff
@@ -1389,7 +1389,11 @@ bool CXXRecordDecl::isLiteral() const {
                              : hasTrivialDestructor()))
     return false;
 
-  if (isLambda() && !LangOpts.CPlusPlus17)
+  // Lambdas are literal types since C++17.
```
Comment (erichkeane) [inline clang/lib/AST/DeclCXX.cpp:1392]: So why is `isAggregate` or `hasConstexprNonCopyMoveConstructor`/etc not just `return true` here?  Am I missing something?

**Review by erichkeane — APPROVED** [review]

**Comment by groundswellaudio** [comment]
Thanks for the fix @cor3ntin. There is still a problem however with `std::optional`, which I think boils down to implicit destructor and anonymous union : 

```cpp
struct A {
  A ();
  ~A();
};

template <class T>
struct opt
{
  union {
    char c;
    T data;
  };
  
  constexpr opt() {}
  
  constexpr ~opt() {
    if (engaged)
      data.~T();
  }
  
  bool engaged = false;
};

consteval void foo()
{
  opt<A> a;
}
```

This fails with `non-literal type 'opt<A>' cannot be used in a constant expression`, but compiles if A has no destructor (or if that destructor is constexpr). 

I suspect that the culprit is here : https://github.com/llvm/llvm-project/blob/fca6992be1f272f5e997bd510ca03c9389550c13/clang/lib/AST/DeclCXX.cpp#L550

The implicit destructor of the anonymous union is not constexpr because one of its subobjects has a non-constexpr destructor.

**Comment by cor3ntin** [comment]
@groundswellaudio Thanks for letting me know. can you create a separate issue ?

**Comment by cor3ntin** [comment]
@groundswellaudio actually, that should be fixed by https://github.com/llvm/llvm-project/pull/77753
@Fznamznon Can you confirm?

**Comment by Fznamznon** [comment]
> @groundswellaudio actually, that should be fixed by https://github.com/llvm/llvm-project/pull/77753
@Fznamznon Can you confirm?

Yes, it does make an error to go away. However the function `foo` from the [code](https://github.com/llvm/llvm-project/pull/78195#issuecomment-1895950521) won't be callable. If it is `constexpr` instead of `consteval`, then it is callable outside of constexpr context.

**Comment by groundswellaudio** [comment]
@Fznamznon 
@cor3ntin 
So https://github.com/llvm/llvm-project/pull/77753 doesn't solve the problem of an anonymous union not being literal when it should be, right?

**Comment by cor3ntin** [comment]
Looking at that again, i suspect I did not handle indirect fields properly. I'll try to get to that. Sorry!

**Comment by cor3ntin** [comment]
> @Fznamznon @cor3ntin So #77753 doesn't solve the problem of the enclosing class of an anonymous union not being literal when it should be, right?

It should actually, CF comments in that PR (I got your code working locally)

**Comment by Fznamznon** [comment]
@cor3ntin , perhaps a dumb question, but https://github.com/llvm/llvm-project/pull/77753 should only help in post C++20 modes since only there destructor can be `constexpr`. How about older versions? Clang will still emit an error that the type from https://github.com/llvm/llvm-project/pull/78195#issuecomment-1895950521 is not literal in C++17 for example.

**Pull request merged**